perf_benchmark/benchmark_main.cc

```diff
@@ -0,0 +1,482 @@
+  if (streaming) {
+    std::string streaming_msg = GetStreamedJson(json_msg, stream_size);
+    is = absl::make_unique<BenchmarkZeroCopyInputStream>(streaming_msg,
+                                                         chunk_per_msg);
+  } else {
+    is = absl::make_unique<BenchmarkZeroCopyInputStream>(std::string(json_msg),
+                                                         chunk_per_msg);
+  }
+
+  // Benchmark the transcoding process
+  std::string message;
+  for (auto s: state) {
+    JsonRequestTranslator translator
+        (type_helper.Resolver(), is.get(), request_info, streaming, false);
+    MessageStream& out = translator.Output();
+
+    if (!out.Status().ok()) {
+      return absl::InternalError(out.Status().ToString());
+    }
+
+    while (out.NextMessage(&message)) {
+    }
+    is->Reset(); // low overhead.
+  }
+
+  // Add custom benchmark counters
+  auto request_processed = static_cast<double>(state.iterations());
+  auto message_processed =
+      static_cast<double>(state.iterations() * (streaming ? stream_size : 1));
+  auto bytes_processed =
+      static_cast<double>(state.iterations() * is->TotalBytes());
+  state.counters["byte_throughput"] = Counter(bytes_processed, Counter::kIsRate,
+                                              Counter::kIs1024);
+  state.counters["byte_latency"] =
+      Counter(bytes_processed,
+              Counter::kIsRate | Counter::kInvert,
+              Counter::kIs1024);
+  state.counters["request_throughput"] =
+      Counter(request_processed, Counter::kIsRate);
+  state.counters["request_latency"] =
+      Counter(request_processed, Counter::kIsRate | Counter::kInvert);
+  state.counters["message_throughput"] =
+      Counter(message_processed, Counter::kIsRate);
+  state.counters["message_latency"] =
+      Counter(message_processed, Counter::kIsRate | Counter::kInvert);
+
+  return message;
+}
+
+//
+// Benchmark variable: JSON body length.
+//
+// Helper function for benchmarking single bytes payload translation from JSON.
+void BM_SinglePayloadFromJson(::benchmark::State& state,
+                              uint64_t payload_length,
+                              bool streaming,
+                              uint64_t stream_size) {
+  std::string msg = GetRandomBytesString(payload_length, true);
+  std::string json_msg = absl::StrFormat(R"({"payload" : "%s"})", msg);
+  absl::StatusOr<std::string> proto_str = BenchmarkJsonTranslation(state,
+                                                                   kBytesPayloadMessageType,
+                                                                   json_msg,
+                                                                   streaming,
+                                                                   stream_size,
+                                                                   1);
+  if (!proto_str.ok()) {
+    state.SkipWithError(proto_str.status().ToString().c_str());
+  } else {
+    // Verification
```

Comment:
Are you sure this actually verifies anything?

When you create `BenchmarkZeroCopyInputStream` in line 118, the `msg` might be copied. Not sure if it's referencing the exact same data anymore

Reply:
Yep, it should be good. `BenchmarkZeroCopyInputStream` takes in `absl::string_view`, and if there's a copy, it will be a deep-copy. Also, the argument passed into `BenchmarkZeroCopyInputStream` is `json_msg` instead of `msg`. `json_msg` is created from `msg` using `absl::StrFormat()`, which should also be a deep-copy.

Reply:
As discussed offline, let's move verification to separate unit test